pysmt/solvers/optimizer.py

```diff
@@ -0,0 +1,243 @@
+        one can move arbitrarily close to the optimum without reching
+        it (e.g. "x > 5" has no minimum for x, only an infimum)
+        """
+        raise NotImplementedError
+
+
+    def pareto_optimize(self, cost_functions, callback=None):
+        """This function is a generator returning *all* the pareto-optimal
+        solutions for the problem of minimizing the `cost_functions`
+        keeping the formulae asserted in this optimizer satisfied.
+
+        `cost_functions` must be a list of terms with integer, real or
+        bit-vector types whose values type has to be minimized
+
+        `callback` is a function taking in input a model object that
+        may be called during the search communicating the current best
+        model. If, when called, the function returns True the search
+        is immediately stopped. This is useful to terminate the search
+        in algorithms where termination is not guaranteed or to stop
+        the search prematurely when a criterion is met.
+
+        This function can raise a PysmtUnboundedOptimizationError if
+        the solver detects that the optimum value is either positive
+        or negative infinite or if there is no optimum value because
+        one can move arbitrarily close to the optimum without reching
+        it (e.g. "x > 5" has no minimum for x, only an infimum)
+
+        """
+        raise NotImplementedError
+
+
+
+class SUAOptimizerMixin(Optimizer):
+    """Optimizer mixin using solving under assumptions"""
+
+    def _lt(self, x, y):
+        otype = self.environment.stc.get_type(x)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type() or otype.is_real_type():
+            return mgr.LT(x, y)
+        elif otype.is_bv_type():
+            return mgr.BVULT(x, y)
+
+    def _le(self, x, y):
+        otype = self.environment.stc.get_type(x)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type() or otype.is_real_type():
+            return mgr.LE(x, y)
+        elif otype.is_bv_type():
+            return mgr.BVULE(x, y)
+
+    def optimize(self, cost_function, initial_cost=None, callback=None):
+        last_model = None
+        keep = True
+
+        initial_assumptions = []
+        if initial_cost is not None:
+            initial_assumptions.append(self._lt(cost_function, initial_cost))
+
+        cost_so_far = None
+        while keep:
+            if cost_so_far is not None:
+                k = self._lt(cost_function, cost_so_far)
+                keep = self.solve(assumptions=[k])
+            else:
+                keep = self.solve(assumptions=initial_assumptions)
+
+            if keep:
+                last_model = self.get_model()
+                cost_so_far = self.get_value(cost_function)
+                if callback is not None:
+                    exit_request = callback(last_model)
+                    if exit_request is True:
+                        break
+        return last_model
```

Comment:
Do we want to return also the cost or only the model? It is easy to obtain the cost given the model, but I was thinking that a return value ```(cost_so_far, last_model)``` might be more intuitive.

Reply:
:+1: